File: template/netwrix-python/function/state_manager.py
"""
State Management Framework for OpenFaaS Connectors

This module provides a framework for managing scan/sync execution states,
including stop/pause/resume operations. Connectors can declare their
supported states and hook into state transition events.

Note: Uses `from __future__ import annotations` for forward type references.

Usage:
    class MyConnectorHandler:
        def __init__(self, context):
            self.state_manager = StateManager(
                context=context,
                supported_states={
                    'stop': True,      # This connector supports stopping
                    'pause': False,    # This connector doesn't support pausing
                    'resume': False    # This connector doesn't support resuming
                }
            )

        def handle(self, event, context):
            # Initialize state monitoring
            self.state_manager.initialize()

            # In main loop:
            if self.state_manager.should_stop():
                self.state_manager.shutdown()
                return result

            # Periodically save progress:
            if self.state_manager.should_checkpoint():
                self.state_manager.save_checkpoint({
                    'progress': current_progress,
                    'timestamp': datetime.now().isoformat()
                })
"""

from __future__ import annotations

import logging
import threading
import time
from typing import Optional, Dict, Any, Callable

# Import here to allow for easier mocking in tests
from function.redis_signal_handler import RedisSignalHandler, ScanControlContext

logger = logging.getLogger(__name__)


class StateManager:
    """
    Manages connector execution states (stop/pause/resume).

    Provides a unified interface for all connectors to:
    - Declare supported states
    - Respond to state change requests
    - Monitor for state transitions
    - Save/restore progress checkpoints
    """

    # Valid state transitions
    VALID_TRANSITIONS = {
        "running": ["stopping", "pausing", "completed", "failed", "stopped"],
        "stopping": ["stopped", "failed"],
        "stopped": [],
        "pausing": ["paused", "failed"],
        "paused": ["resuming", "failed"],
        "resuming": ["running", "failed"],
        "completed": [],
        "failed": [],
    }

    # Default supported states (all connectors can support stop)
    DEFAULT_SUPPORTED_STATES = {
        "stop": True,  # Halt execution
        "pause": False,  # Suspend and save state
        "resume": False,  # Continue from pause
    }

    def __init__(
        self,
        context,
        supported_states: Optional[Dict[str, bool]] = None,
        checkpoint_interval: int = 60,
        signal_check_interval: int = 5,
    ):
        """
        Initialize state manager

        Args:
            context: OpenFaaS context object
            supported_states: Dict of {state: bool} indicating support
            checkpoint_interval: Seconds between checkpoints (default 60)
            signal_check_interval: Seconds between signal checks (default 5)
        """
        self.context = context
        self.supported_states = {**self.DEFAULT_SUPPORTED_STATES, **(supported_states or {})}
        self.checkpoint_interval = checkpoint_interval
        self.signal_check_interval = signal_check_interval

        self.current_state = "running"
        self.requested_state = None
        self.redis_handler = None
        self.control_context = None
        self.last_signal_check = time.time()
        self.last_checkpoint = time.time()
        self._state_lock = threading.Lock()
        self._shutdown_event = threading.Event()
        self._on_state_change_callbacks = []

    def initialize(self) -> bool:
        """
        Initialize Redis signal monitoring

        Returns:
            True if initialization successful, False otherwise
        """
        try:
            # Get execution ID
            execution_id = self.context.scan_execution_id
            if not execution_id:
                logger.warning("No execution ID available, signal monitoring disabled")
                return False

            # Initialize Redis handler
            self.redis_handler = RedisSignalHandler()

            if not self.redis_handler.health_check():
                logger.warning("Redis unavailable, signal monitoring disabled")
                return False

            # Create control context
            self.control_context = ScanControlContext(execution_id, self.redis_handler)

            # Update status
            self.redis_handler.update_status(execution_id, "running")

            logger.info(f"State manager initialized (supported_states={self.supported_states})")
            return True

        except Exception as e:
            logger.error(f"Failed to initialize state manager: {e}")
            return False

    def check_for_state_changes(self) -> bool:
        """
        Check for state change requests from Redis

        Returns:
            True if stop was requested, False otherwise
        """
        if not self.control_context:
            return False

        current_time = time.time()
        if current_time - self.last_signal_check < self.signal_check_interval:
            logger.info("check_for_state_changes: too soon")
            return False  # Not time to check yet

        self.last_signal_check = current_time

        # Check for incoming signals
        try:
            # Socket timeout is configured on Redis client prevent indefinite blocking if it becomes unresponsive.
            signal = self.control_context.check_for_signals()
            if signal and self.control_context.stop_requested:
                self.requested_state = "stop"
                # Actually transition to stopping state
                if self.current_state == "running":
                    if self.set_state("stopping"):
                        logger.info("State transitioned (from_state=running, to_state=stopping)")
                return True
        except Exception as e:
            # just return and allow subsequent calls, in case the issue is transient
            logger.warning(f"Error checking state changes: {e}")

        return False

    def should_stop(self) -> bool:
        """
        Check if execution should stop

        Returns:
            True if stop was requested, False otherwise
        """
        self.check_for_state_changes()
        with self._state_lock:
            return self.requested_state == "stop"

    def should_pause(self) -> bool:
        """
        Check if execution should pause

        Returns:
            True if pause was requested and supported, False otherwise
        """
        if not self.supported_states.get("pause", False):
            return False

        self.check_for_state_changes()
        with self._state_lock:
            return (
                self.requested_state == "pause"
                and self.control_context is not None
                and self.control_context.pause_requested
            )

    def should_checkpoint(self) -> bool:
        """
        Check if it's time to save a checkpoint

        Returns:
            True if checkpoint interval has elapsed, False otherwise
        """
        current_time = time.time()
        elapsed = current_time - self.last_checkpoint
        if elapsed >= self.checkpoint_interval:
            return True
        return False

    def save_checkpoint(self, checkpoint_data: Dict[str, Any]) -> Optional[str]:
        """
        Save execution progress checkpoint

        Args:
            checkpoint_data: Dictionary with checkpoint information
                - progress: current progress data
                - timestamp: ISO8601 timestamp
                - additional fields as needed

        Returns:
            Checkpoint ID if successful, None otherwise
        """
        if not self.redis_handler or not self.control_context:
            return None

        try:
            execution_id = self.context.scan_execution_id
            checkpoint_id = self.redis_handler.save_checkpoint(execution_id, checkpoint_data)
            self.last_checkpoint = time.time()

            logger.debug(f"Checkpoint saved (execution_id={execution_id}, checkpoint_id={checkpoint_id})")
            return checkpoint_id

        except Exception as e:
            logger.warning(f"Failed to save checkpoint: {e}")
            return None

    def set_state(self, new_state: str) -> bool:
        """
        Transition to a new state

        Args:
            new_state: Target state name

        Returns:
            True if transition is valid and successful, False otherwise
        """
        with self._state_lock:
            valid_transitions = self.VALID_TRANSITIONS.get(self.current_state, [])

            if new_state not in valid_transitions:
                logger.warning(f"Invalid state transition (from_state={self.current_state}, to_state={new_state})")
                return False

            old_state = self.current_state
            self.current_state = new_state

            logger.info(f"State transitioned (from_state={old_state}, to_state={new_state})")

        # Call callbacks outside lock to avoid deadlocks
        self._trigger_state_change_callbacks(old_state, new_state)
        return True

    def on_state_change(self, callback: Callable[[str, str], None]) -> None:
        """
        Register callback for state change events

        Args:
            callback: Function(old_state, new_state) to call on state changes
        """
        self._on_state_change_callbacks.append(callback)

    def _trigger_state_change_callbacks(self, old_state: str, new_state: str):
        """Trigger all registered callbacks for state change"""
        for callback in self._on_state_change_callbacks:
            try:
                callback(old_state, new_state)
            except Exception as e:
                logger.error(f"Error in state change callback: {e}")

    def shutdown(self, final_status: str = "stopped") -> bool:
        """
        Gracefully shut down and transition to final state

        Args:
            final_status: Final state ('stopped', 'completed', 'failed')

        Returns:
            True if shutdown successful, False otherwise
        """
        try:
            # Transition state
            if not self.set_state(final_status):
                logger.warning(f"Could not transition to {final_status}")
                return False

            # Update Redis status
            if self.redis_handler:
                execution_id = self.context.scan_execution_id
                self.redis_handler.update_status(
                    execution_id, final_status, "Execution stopped", {"partial_data": final_status == "stopped"}
                )

                # Cleanup streams
                self.redis_handler.cleanup_streams(execution_id)

            self._shutdown_event.set()
            logger.info(f"State manager shutdown with final status: {final_status}")
            return True

        except Exception as e:
            logger.error(f"Error during shutdown: {e}")
            return False

    def is_shutdown(self) -> bool:
        """Check if shutdown has been initiated"""
        return self._shutdown_event.is_set()

    def supports_state(self, state: str) -> bool:
        """Check if connector supports a state"""
        return self.supported_states.get(state, False)

    def get_supported_states(self) -> Dict[str, bool]:
        """Get all supported states"""
        return self.supported_states.copy()

    def get_current_state(self) -> str:
        """Get current execution state"""
        with self._state_lock:
            return self.current_state

    def close(self):
        """Clean up resources"""
        if self.redis_handler:
            self.redis_handler.close()

File: template/netwrix-csharp/ConnectorFramework/StateManager.cs
namespace Netwrix.ConnectorFramework;
/// <summary>
/// Manages scan execution states (running/stopping/stopped/pausing/paused/resuming/completed/failed).
/// Uses SemaphoreSlim for all state mutations — safe to await inside async methods.
/// </summary>
public sealed class StateManager : IAsyncDisposable
{
    private static readonly IReadOnlyDictionary<string, IReadOnlyList<string>> ValidTransitions =
        new Dictionary<string, IReadOnlyList<string>>
        {
            [ScanStatus.Running] = [ScanStatus.Stopping, ScanStatus.Pausing, ScanStatus.Completed, ScanStatus.CompletedWithErrors, ScanStatus.Failed],
            [ScanStatus.Stopping] = [ScanStatus.Stopped, ScanStatus.Failed],
            [ScanStatus.Stopped] = [],
            [ScanStatus.Pausing] = [ScanStatus.Paused, ScanStatus.Failed],
            [ScanStatus.Paused] = [ScanStatus.Resuming, ScanStatus.Failed, ScanStatus.Stopped],
            [ScanStatus.Resuming] = [ScanStatus.Running, ScanStatus.Failed],
            [ScanStatus.Completed] = [],
            [ScanStatus.CompletedWithErrors] = [],
            [ScanStatus.Failed] = [],
        };

    private readonly RedisSignalHandler _redis;
    private readonly ScanShutdownService _shutdown;
    private readonly IScanProgress _progress;
    private readonly ILogger<StateManager> _logger;
    private readonly SemaphoreSlim _semaphore = new(1, 1);
    private readonly List<Func<string, string, Task>> _callbacks = [];

    private volatile string _currentState = ScanStatus.Running;
    private string? _requestedState;
    private DateTimeOffset _lastSignalCheck = DateTimeOffset.MinValue;

    public SupportedStates SupportedStates { get; }
    public int SignalCheckIntervalSeconds { get; }

    /// <summary>
    /// A CancellationToken that is cancelled when a Redis STOP signal arrives or ShutdownAsync is called.
    /// Backed by the singleton <see cref="ScanShutdownService"/> so all DI scopes observe the same signal.
    /// Combine with Kestrel's token using CancellationTokenSource.CreateLinkedTokenSource.
    /// </summary>
    public CancellationToken Token => _shutdown.Token;

    public StateManager(
        RedisSignalHandler redis,
        ScanShutdownService shutdown,
        IScanProgress progress,
        ILogger<StateManager> logger,
        SupportedStates? supportedStates = null,
        int signalCheckIntervalSeconds = 5)
    {
        _redis = redis;
        _shutdown = shutdown;
        _progress = progress;
        _logger = logger;
        SupportedStates = supportedStates ?? new SupportedStates();
        SignalCheckIntervalSeconds = signalCheckIntervalSeconds;
    }

    // ── State queries ─────────────────────────────────────────────────────────

    /// <summary>
    /// Returns the current state. Safe to read from any thread — backed by a volatile field.
    /// Writes still go through the semaphore to enforce valid transition logic.
    /// </summary>
    public string CurrentState => _currentState;

    public bool IsShutdown => _shutdown.IsCancellationRequested;

    // ── Signal monitoring ─────────────────────────────────────────────────────

    /// <summary>
    /// Polls Redis for control signals if the check interval has elapsed.
    /// Returns true if a STOP signal was received.
    /// </summary>
    public async Task<bool> ShouldStopAsync(string executionId, CancellationToken ct = default)
    {
        if (!SupportedStates.Stop)
        {
            return false;
        }

        await CheckForSignalsAsync(executionId, ct);

        await _semaphore.WaitAsync(ct);
        try { return _requestedState == "stop"; }
        finally { _semaphore.Release(); }
    }

    /// <summary>
    /// Polls Redis for control signals if the check interval has elapsed.
    /// Returns true if a PAUSE signal was received.
    /// </summary>
    public async Task<bool> ShouldPauseAsync(string executionId, CancellationToken ct = default)
    {
        if (!SupportedStates.Pause)
        {
            return false;
        }

        await CheckForSignalsAsync(executionId, ct);

        await _semaphore.WaitAsync(ct);
        try { return _requestedState == "pause"; }
        finally { _semaphore.Release(); }
    }

    private async Task CheckForSignalsAsync(string executionId, CancellationToken ct)
    {
        var now = DateTimeOffset.UtcNow;

        // Read and update atomically under the semaphore to prevent two concurrent callers
        // from both passing the interval check before either records that a check is in progress.
        await _semaphore.WaitAsync(ct);
        var shouldCheck = (now - _lastSignalCheck).TotalSeconds >= SignalCheckIntervalSeconds;
        if (shouldCheck)
        {
            _lastSignalCheck = now;
        }

        _semaphore.Release();

        if (!shouldCheck)
        {
            return;
        }

        try
        {
            var signal = await _redis.CheckControlSignalAsync(executionId, ct: ct);
            if (signal is null)
            {
                return;
            }

            var (action, _) = signal.Value;

            if (action == "STOP")
            {
                await SetStateInternalAsync("stopping", ct);
                await _semaphore.WaitAsync(ct);
                try { _requestedState = "stop"; }
                finally { _semaphore.Release(); }
                _logger.LogInformation("STOP signal received for execution {ExecutionId}", executionId);
            }
            else if (action == "PAUSE" && SupportedStates.Pause)
            {
                await SetStateInternalAsync("pausing", ct);
                await _semaphore.WaitAsync(ct);
                try { _requestedState = "pause"; }
                finally { _semaphore.Release(); }
                _logger.LogInformation("PAUSE signal received for execution {ExecutionId}", executionId);
            }
            else if (action == "RESUME" && SupportedStates.Resume)
            {
                await _semaphore.WaitAsync(ct);
                try { _requestedState = null; }
                finally { _semaphore.Release(); }
                _logger.LogInformation("RESUME signal received for execution {ExecutionId}", executionId);
            }
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Error checking signals for execution {ExecutionId}", executionId);
        }
    }

    // ── State transitions ─────────────────────────────────────────────────────

    /// <summary>
    /// Attempts to transition to <paramref name="newState"/>.
    /// Returns false if the transition is invalid from the current state.
    /// </summary>
    public async Task<bool> SetStateAsync(string newState, CancellationToken ct = default)
        => await SetStateInternalAsync(newState, ct);

    private async Task<bool> SetStateInternalAsync(string newState, CancellationToken ct)
    {
        string oldState;

        await _semaphore.WaitAsync(ct);
        try
        {
            if (newState == _currentState)
            {
                return true;
            }

            if (!ValidTransitions.TryGetValue(_currentState, out var allowed) || !allowed.Contains(newState))
            {
                _logger.LogWarning("Invalid state transition from {OldState} to {NewState}", _currentState, newState);
                return false;
            }

            oldState = _currentState;
            _currentState = newState;
            _logger.LogInformation("State transitioned from {OldState} to {NewState}", oldState, newState);
        }
        finally
        {
            _semaphore.Release();
        }

        // Fire callbacks outside the semaphore to avoid deadlocks
        await FireCallbacksAsync(oldState, newState);
        return true;
    }

    // ── Shutdown ──────────────────────────────────────────────────────────────

    /// <summary>
    /// Transitions to <paramref name="finalStatus"/>, publishes the status to Redis, and cancels the token.
    /// </summary>
    public async Task<bool> ShutdownAsync(
        string executionId,
        string finalStatus = ScanStatus.Stopped,
        CancellationToken ct = default)
    {
        try
        {
            if (!await SetStateInternalAsync(finalStatus, ct))
            {
                _logger.LogWarning("Could not transition to final state {FinalStatus}", finalStatus);
                return false;
            }

            await _progress.UpdateExecutionAsync(
                status: finalStatus,
                completedAt: finalStatus != ScanStatus.Paused ? DateTimeOffset.UtcNow : null,
                ct: ct);

            await _redis.UpdateStatusAsync(
                executionId,
                finalStatus,
                "Execution stopped",
                partialData: finalStatus == ScanStatus.Stopped,
                ct: ct);

            await _redis.CleanupStreamsAsync(executionId, ct);

            _shutdown.Cancel();
            _logger.LogInformation("StateManager shutdown with status {FinalStatus}", finalStatus);
            return true;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error during StateManager shutdown");
            return false;
        }
    }

    // ── Callbacks ─────────────────────────────────────────────────────────────

    /// <summary>
    /// Registers an async callback invoked after every successful state transition.
    /// </summary>
    public void OnStateChange(Func<string, string, Task> callback)
        => _callbacks.Add(callback);

    private async Task FireCallbacksAsync(string oldState, string newState)
    {
        foreach (var cb in _callbacks)
        {
            try { await cb(oldState, newState); }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error in state change callback");
            }
        }
    }

    public ValueTask DisposeAsync()
    {
        _semaphore.Dispose();
        return ValueTask.CompletedTask;
    }
}

/// <summary>
/// Declares which control operations this connector supports.
/// </summary>
public record SupportedStates(bool Stop = true, bool Pause = true, bool Resume = false);
